### PR TITLE
Add an option in the Anvil DSL to sync generated sources in the IDE.

### PR DESCRIPTION
**Pull request by vRallev**
https://user-images.githubusercontent.com/1558045/140219280-ec7eb47b-6fe7-4a7a-9c0a-f92e4cf1d98c.mov

### CODE REVIEW

**Review by christiandeange — APPROVED** [review]

File: gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt [hunk 1]
```diff
@@ -130,6 +130,22 @@ internal open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
...
+    if (variant.variantFilter.syncGeneratedSources) {
+      val isIdeSyncProvider = project.providers
+        .systemProperty("idea.sync.active")
+        .forUseAtConfigurationTime()
+
+      if (isIdeSyncProvider.getOrElse("false").toBoolean()) {
+        // Only add source sets during the IDE sync. Don't add them for compilation, otherwise
+        // we'll see weird compile errors especially with incremental compilation. For a longer
+        // explanation why this is a bad idea, see here:
+        // https://github.com/square/anvil/pull/207#issuecomment-850768750
```
Comment (autonomousapps) [inline gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt:142]: Super interesting. I was wondering why these wouldn't get added to the normal compilation source sets.

File: gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt [hunk 2]
```diff
@@ -130,6 +130,22 @@ internal open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
       "anvil${File.separator}src-gen-${variant.name}"
     )
 
+    if (variant.variantFilter.syncGeneratedSources) {
+      val isIdeSyncProvider = project.providers
+        .systemProperty("idea.sync.active")
+        .forUseAtConfigurationTime()
+
+      if (isIdeSyncProvider.getOrElse("false").toBoolean()) {
```
Comment (autonomousapps) [inline gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt:138]: From a CC perspective, this all looks fine.

**Review by autonomousapps — APPROVED** [review]

File: gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt (hunk 1 above)
Comment (vRallev) [reply gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt:142]: Compiler plugins aren't ready for prime time, yet :-)

File: gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt (hunk 2 above)
Comment (vRallev) [reply gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt:138]: Thanks!

**Comment by vRallev** [comment]
Thanks for the reviews!

**Pull request merged**